### DISCUSSION TITLE
turning off peripherals before idle?

### DISCUSSION THREAD

**Discussion by arthurv** [jefftenney/LPTIM-Tick#4] [Q&A]
It works great! However, my application uses SPI2 and as soon as I turn it on in CubeMX consumption jumps up to 660 uA in test 1.

Is there a place in the code to turn off peripherals like SPI2 before entering idle to keep consumption low?

**Comment by jefftenney** [comment]
In ultra-low power applications, I think the best strategy is for the task that uses a device to disable it when finished.  

When a task wants to use SPI2, it would:

1. Tell the power manager that SPI2 is in use
2. Enable SPI2
3. Use SPI2
4. Disable SPI2
5. Tell the power manager that SPI2 is not in use

This repo uses ulp.c/.h as the power manager.  In your case, you would add `ulpPERIPHERAL_SPI2` to the list of peripherals [here](https://github.com/jefftenney/LPTIM-Tick/blob/e3c71f627d62e1742cc99669dc508dadbe0580e9/Core/Inc/ulp.h#L25). 

With this strategy, power consumption will automatically be ultra low whenever possible.
